File: android/app/src/main/java/com/example/bonermohis/data/ElectricityRepository.kt
package com.example.bonermohis.data

import kotlinx.coroutines.flow.Flow
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.random.Random

class ElectricityRepository(
    private val accountDao: AccountDao,
    private val dailyUsageDao: DailyUsageDao
) {
    val allAccounts: Flow<List<Account>> = accountDao.getAllFlow()

    fun getHistoryFlow(accountId: Int): Flow<List<DailyUsageHistory>> {
        return dailyUsageDao.getHistoryForAccountFlow(accountId)
    }

    suspend fun insertAccount(account: Account): Long {
        return accountDao.insert(account)
    }

    suspend fun deleteAccount(accountId: Int) {
        val account = accountDao.getById(accountId)
        if (account != null) {
            dailyUsageDao.deleteHistoryForAccount(accountId)
            accountDao.delete(account)
        }
    }

    suspend fun topUp(accountId: Int, amount: Double) {
        val account = accountDao.getById(accountId) ?: return
        val newBalance = account.balance + amount
        val updated = account.copy(
            balance = newBalance,
            lastUpdated = System.currentTimeMillis()
        )
        accountDao.update(updated)
    }

    suspend fun resetCycle(accountId: Int) {
        val account = accountDao.getById(accountId) ?: return
        val updated = account.copy(
            monthlyKwh = 0.0,
            currentSlab = 0,
            slabUsage = 0.0,
            lastUpdated = System.currentTimeMillis()
        )
        accountDao.update(updated)
    }

    suspend fun simulateDay(accountId: Int, customKwh: Double? = null) {
        val account = accountDao.getById(accountId) ?: return
        
        // 1. Generate usage
        val kwhUsed = customKwh ?: (4.0 + Random.nextDouble() * 8.0)
        
        // 2. Add to monthly total usage
        val oldMonthlyKwh = account.monthlyKwh
        val newMonthlyKwh = oldMonthlyKwh + kwhUsed
        
        // 3. Compute cost
        val costOfNewTotal = CalculationsHelper.calculateCost(newMonthlyKwh, account.distributor)
        val costOfOldTotal = CalculationsHelper.calculateCost(oldMonthlyKwh, account.distributor)
        val dailyCost = maxOf(0.0, costOfNewTotal - costOfOldTotal)
        
        // 4. Deduct cost from balance
        val newBalance = maxOf(0.0, account.balance - dailyCost)
        
        // 5. Determine active slab tier stats
        val slabStats = CalculationsHelper.getSlabDetails(newMonthlyKwh, account.distributor)
        
        // 6. Write daily history record
        dailyUsageDao.insert(
            DailyUsageHistory(
                accountId = accountId,
                dateEpoch = System.currentTimeMillis(),
                consumptionKwh = kwhUsed,
                cost = dailyCost
            )
        )
        
        // 7. Update account metrics
        val updated = account.copy(
            balance = newBalance,
            monthlyKwh = newMonthlyKwh,
            yesterdayUsage = dailyCost,
            currentSlab = slabStats.index,
            slabUsage = newMonthlyKwh - slabStats.slabMin,
            lastUpdated = System.currentTimeMillis()
        )
        
        accountDao.update(updated)
    }

    suspend fun syncAccount(accountId: Int) = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val account = accountDao.getById(accountId) ?: return@withContext
        if (account.distributor != "desco") return@withContext

        try {
            // 1. Fetch balance
            val balanceUrl = "https://prepaid.desco.org.bd/api/tkdes/customer/getBalance?accountNo=${account.accountNo}&meterNo=${account.meterNo}"
            val balanceRes = httpGet(balanceUrl) ?: throw Exception("Failed to fetch DESCO balance")
            val balanceObj = org.json.JSONObject(balanceRes)
            if (balanceObj.getInt("code") != 200) throw Exception(balanceObj.optString("desc", "DESCO error"))
            
            val data = balanceObj.getJSONObject("data")
            val liveBalance = data.getDouble("balance")
            // NOTE: currentMonthConsumption is BDT cost, NOT kWh.
            // The actual kWh is derived from consumedUnit in the daily consumption endpoint below.
            var liveMonthlyKwh = 0.0

            // 2. Fetch daily consumption for last 15 days
            val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val dateTo = sdf.format(Date())
            val cal = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, -15) }
            val dateFrom = sdf.format(cal.time)

            val consumptionUrl = "https://prepaid.desco.org.bd/api/tkdes/customer/getCustomerDailyConsumption?accountNo=${account.accountNo}&meterNo=${account.meterNo}&dateFrom=$dateFrom&dateTo=$dateTo"
            val consumptionRes = httpGet(consumptionUrl)
            var yesterdayCost = account.yesterdayUsage

            if (consumptionRes != null) {
                val consumptionObj = org.json.JSONObject(consumptionRes)
                if (consumptionObj.getInt("code") == 200) {
                    val dataArray = consumptionObj.getJSONArray("data")
                    val list = mutableListOf<org.json.JSONObject>()
                    for (i in 0 until dataArray.length()) {
                        list.add(dataArray.getJSONObject(i))
                    }
                    
                    list.sortBy { it.getString("date") }
                    
                    // Derive monthly kWh from consumedUnit (cumulative all-time meter reading).
                    // consumedTaka resets at month boundary; consumedUnit never resets.
                    // Base must be the LAST entry from the PREVIOUS month so that the
                    // first day of the current month is included in the delta.
                    if (list.isNotEmpty()) {
                        val latestEntry = list.last()
                        val currentMonthStr = latestEntry.getString("date").substring(0, 7) // "yyyy-MM"
                        val lastOfPrevMonth = list.lastOrNull { !it.getString("date").startsWith(currentMonthStr) }
                        val baseUnit = lastOfPrevMonth?.getDouble("consumedUnit")
                            ?: list.firstOrNull { it.getString("date").startsWith(currentMonthStr) }?.getDouble("consumedUnit")
                            ?: latestEntry.getDouble("consumedUnit")
                        liveMonthlyKwh = maxOf(0.0, latestEntry.getDouble("consumedUnit") - baseUnit)
                        liveMonthlyKwh = Math.round(liveMonthlyKwh * 1000) / 1000.0
                    }
                    
                    for (i in 0 until list.size) {
                        val current = list[i]
                        val prev = if (i > 0) list[i - 1] else null
                        
                        var dailyKwh = 0.0
                        var dailyCost = 0.0
                        
                        if (prev != null) {
                            dailyKwh = current.getDouble("consumedUnit") - prev.getDouble("consumedUnit")
                            
                            val currentMonth = current.getString("date").split("-")[1]
                            val prevMonth = prev.getString("date").split("-")[1]
                            if (currentMonth == prevMonth) {
                                dailyCost = maxOf(0.0, current.getDouble("consumedTaka") - prev.getDouble("consumedTaka"))
                            } else {
                                dailyCost = current.getDouble("consumedTaka")
                            }
                        }

                        if (dailyKwh > 0.0 || dailyCost > 0.0) {
                            val dateEpoch = sdf.parse(current.getString("date"))?.time ?: System.currentTimeMillis()
                            val existing = dailyUsageDao.getRecordForDay(accountId, dateEpoch)
                            if (existing != null) {
                                dailyUsageDao.update(existing.copy(consumptionKwh = dailyKwh, cost = dailyCost))
                            } else {
                                dailyUsageDao.insert(
                                    DailyUsageHistory(
                                        accountId = accountId,
                                        dateEpoch = dateEpoch,
                                        consumptionKwh = dailyKwh,
                                        cost = dailyCost
                                    )
                                )
                            }
                        }
                    }

                    if (list.size >= 2) {
                        val last = list[list.size - 1]
                        val secLast = list[list.size - 2]
                        val lastMonth = last.getString("date").split("-")[1]
                        val secLastMonth = secLast.getString("date").split("-")[1]
                        if (lastMonth == secLastMonth) {
                            yesterdayCost = maxOf(0.0, last.getDouble("consumedTaka") - secLast.getDouble("consumedTaka"))
                        } else {
                            yesterdayCost = last.getDouble("consumedTaka")
                        }
                    }
                }
            }

            // 3. Compute active slab details
            val slabStats = CalculationsHelper.getSlabDetails(liveMonthlyKwh, "desco")

            // 4. Update and save account
            val updated = account.copy(
                balance = liveBalance,
                monthlyKwh = liveMonthlyKwh,
                yesterdayUsage = yesterdayCost,
                currentSlab = slabStats.index,
                slabUsage = liveMonthlyKwh - slabStats.slabMin,
                lastUpdated = System.currentTimeMillis()
            )
            accountDao.update(updated)

        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun httpGet(urlString: String): String? {
        var connection: java.net.HttpURLConnection? = null
        return try {
            val url = java.net.URL(urlString)
            connection = url.openConnection() as java.net.HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = 10000
            connection.readTimeout = 10000
            connection.inputStream.bufferedReader().use { it.readText() }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        } finally {
            connection?.disconnect()
        }
    }
}

File: lib/data/repository/electricity_repository.dart
import 'dart:convert';
import 'dart:math';
import 'package:intl/intl.dart';
import '../models/account.dart';
import '../models/daily_usage_history.dart';
import '../database/account_dao.dart';
import '../database/daily_usage_dao.dart';
import '../calculations_helper.dart';
import '../desco_http_client.dart';
import '../../background/notification_setup.dart';

class ElectricityRepository {
  final AccountDao accountDao;
  final DailyUsageDao dailyUsageDao;

  ElectricityRepository({
    required this.accountDao,
    required this.dailyUsageDao,
  });

  Stream<List<Account>> get allAccounts => accountDao.watchAll();

  Stream<List<DailyUsageHistory>> historyStream(int accountId) =>
      dailyUsageDao.watchHistoryForAccount(accountId);

  Future<int> insertAccount(Account account) => accountDao.insert(account);

  Future<void> deleteAccount(int accountId) async {
    final account = await accountDao.getById(accountId);
    if (account == null) return;
    await dailyUsageDao.deleteHistoryForAccount(accountId);
    await accountDao.delete(account);
  }

  Future<String?> syncAccount(int accountId) async {
    final account = await accountDao.getById(accountId);
    if (account == null) return 'Account not found';
    if (account.distributor != 'desco') return null;

    try {
      final client = await createDescoHttpClient();
      try {
        final balanceUri = Uri.parse(
          'https://prepaid.desco.org.bd/api/tkdes/customer/getBalance'
          '?accountNo=${account.accountNo}&meterNo=${account.meterNo}',
        );
        final balanceRes = await client
            .get(balanceUri)
            .timeout(const Duration(seconds: 10));
        final balanceObj = jsonDecode(balanceRes.body) as Map<String, dynamic>;
        if ((balanceObj['code'] as int) != 200) {
          return balanceObj['desc']?.toString() ?? 'DESCO API error';
        }
        final data = balanceObj['data'] as Map<String, dynamic>;
        double numToDouble(dynamic val) {
          if (val == null) return 0.0;
          if (val is num) return val.toDouble();
          return double.tryParse(val.toString()) ?? 0.0;
        }

        final liveBalance = numToDouble(data['balance']);

        final sdf = DateFormat('yyyy-MM-dd');
        final dateTo = sdf.format(DateTime.now());
        final dateFrom = sdf.format(
          DateTime.now().subtract(const Duration(days: 35)),
        );

        final consumptionUri = Uri.parse(
          'https://prepaid.desco.org.bd/api/tkdes/customer/getCustomerDailyConsumption'
          '?accountNo=${account.accountNo}&meterNo=${account.meterNo}'
          '&dateFrom=$dateFrom&dateTo=$dateTo',
        );

        double yesterdayCost = account.yesterdayUsage;
        double liveMonthlyKwh = 0.0;
        double baseUnit = 0.0;

        try {
          final consumptionRes = await client
              .get(consumptionUri)
              .timeout(const Duration(seconds: 10));
          final consumptionObj =
              jsonDecode(consumptionRes.body) as Map<String, dynamic>;

          if ((consumptionObj['code'] as int) == 200) {
            final dataArray = (consumptionObj['data'] as List)
                .cast<Map<String, dynamic>>();
            final list = List<Map<String, dynamic>>.from(dataArray)
              ..sort(
                (a, b) => (a['date'] as String).compareTo(b['date'] as String),
              );

            if (list.isNotEmpty) {
              final latestEntry = list.last;
              final currentMonthStr = (latestEntry['date'] as String).substring(
                0,
                7,
              );
              final lastOfPrevMonth = list.lastWhere(
                (e) => !(e['date'] as String).startsWith(currentMonthStr),
                orElse: () => <String, dynamic>{},
              );
              baseUnit = lastOfPrevMonth.isNotEmpty
                  ? numToDouble(lastOfPrevMonth['consumedUnit'])
                  : numToDouble(
                      list.firstWhere(
                        (e) => (e['date'] as String).startsWith(currentMonthStr),
                        orElse: () => latestEntry,
                      )['consumedUnit'],
                    );

              liveMonthlyKwh = max(
                0.0,
                numToDouble(latestEntry['consumedUnit']) - baseUnit,
              );
              liveMonthlyKwh = (liveMonthlyKwh * 1000).roundToDouble() / 1000.0;
            }

            for (int i = 0; i < list.length; i++) {
              final current = list[i];
              final prev = i > 0 ? list[i - 1] : null;

              double dailyKwh = 0.0;
              double dailyCost = 0.0;

              if (prev != null) {
                final curUnit = numToDouble(current['consumedUnit']);
                final prevUnit = numToDouble(prev['consumedUnit']);
                final curTaka = numToDouble(current['consumedTaka']);
                final prevTaka = numToDouble(prev['consumedTaka']);

                dailyKwh = max(0.0, curUnit - prevUnit);
                final curMonth = (current['date'] as String).split('-')[1];
                final prevMonth = (prev['date'] as String).split('-')[1];
                if (curMonth == prevMonth) {
                  dailyCost = max(0.0, curTaka - prevTaka);
                } else {
                  dailyCost = curTaka;
                }

                // Fallback: If odometer reading did not update (dailyKwh is 0) but cost was billed,
                // estimate daily kWh using active slab rate.
                if (dailyKwh == 0.0 && dailyCost > 5.0) {
                  final curUnitVal = numToDouble(current['consumedUnit']);
                  final monthlyKwhAtDate = curUnitVal > 0.0 ? curUnitVal - baseUnit : liveMonthlyKwh;
                  final slabDetails = CalculationsHelper.getSlabDetails(
                    max(0.0, monthlyKwhAtDate),
                    provider: 'desco',
                  );
                  final rate = slabDetails.rate > 0 ? slabDetails.rate : 6.18;
                  dailyKwh = (dailyCost / rate * 1000).roundToDouble() / 1000.0;
                }
              }

              if (dailyKwh > 0.0 || dailyCost > 0.0) {
                final dateEpoch = sdf
                    .parse(current['date'] as String)
                    .millisecondsSinceEpoch;
                final existing = await dailyUsageDao.getRecordForDay(
                  accountId,
                  dateEpoch,
                );
                if (existing != null) {
                  await dailyUsageDao.update(
                    existing.copyWith(
                      consumptionKwh: dailyKwh,
                      cost: dailyCost,
                    ),
                  );
                } else {
                  await dailyUsageDao.insert(
                    DailyUsageHistory(
                      accountId: accountId,
                      dateEpoch: dateEpoch,
                      consumptionKwh: dailyKwh,
                      cost: dailyCost,
                    ),
                  );
                }
              }
            }

            if (list.length >= 2) {
              final last = list[list.length - 1];
              final secLast = list[list.length - 2];
              final lastMonth = (last['date'] as String).split('-')[1];
              final secLastMonth = (secLast['date'] as String).split('-')[1];
              if (lastMonth == secLastMonth) {
                yesterdayCost = max(
                  0.0,
                  numToDouble(last['consumedTaka']) -
                      numToDouble(secLast['consumedTaka']),
                );
              } else {
                yesterdayCost = numToDouble(last['consumedTaka']);
              }
            }
          }
        } catch (_) {}

        final slabStats = CalculationsHelper.getSlabDetails(
          liveMonthlyKwh,
          provider: 'desco',
        );

        await accountDao.update(
          account.copyWith(
            balance: liveBalance,
            monthlyKwh: liveMonthlyKwh,
            yesterdayUsage: yesterdayCost,
            currentSlab: slabStats.index,
            slabUsage: liveMonthlyKwh - slabStats.slabMin,
            lastUpdated: DateTime.now().millisecondsSinceEpoch,
          ),
        );

        // Notify immediately after sync if balance is critically low.
        await checkAndNotifyForAccount(
          id: accountId,
          nickname: account.nickname,
          balance: liveBalance,
          yesterdayUsage: yesterdayCost,
        );

        return null;
      } finally {
        client.close();
      }
    } catch (e) {
      return e.toString();
    }
  }

  Future<void> syncAllAccounts() async {
    final accounts = await accountDao.getAll();
    for (final account in accounts) {
      if (account.id != null && account.distributor == 'desco') {
        await syncAccount(account.id!);
      }
    }
  }
}

